hwdec/opengl: Add support for CUDA and cuvid/NvDecode

philipl opened the pull request:
I agree that my changes can be relicensed to LGPL 2.1 or later.

Nvidia's "NvDecode" API (up until recently called "cuvid" is a cross
platform, but nvidia proprietary API that exposes their hardware
video decoding capabilities. It is analogous to their DXVA or VDPAU
support on Windows or Linux but without using platform specific API
calls.

As a rule, you'd rather use DXVA or VDPAU as these are more mature
and well supported APIs, but on Linux, VDPAU is falling behind the
hardware capabilities, and there's no sign that nvidia are making
the investments to update it.

Most concretely, this means that there is no VP8/9 or HEVC Main10
support in VDPAU. On the other hand, NvDecode does export vp8/9 and
partial support for HEVC Main10 (more on that below).

ffmpeg already has support in the form of the "cuvid" family of
decoders. Due to the design of the API, it is best exposed as a full
decoder rather than an hwaccel. As such, there are decoders like
h264_cuvid, hevc_cuvid, etc.

These decoders support two output paths today - in both cases, NV12
frames are returned, either in CUDA device memory or regular system
memory.

In the case of the system memory path, the decoders can be used
as-is in mpv today with a command line like:

mpv --vd=lavc:h264_cuvid foobar.mp4

Doing this will take advantage of hardware decoding, but the cost
of the memcpy to system memory adds up, especially for high
resolution video (4K etc).

To avoid that, we need an hwdec that takes advantage of CUDA's
OpenGL interop to copy from device memory into OpenGL textures.

That is what this change implements.

The process is relatively simple as only basic device context
aquisition needs to be done by us - the CUDA buffer pool is managed
by the decoder - thankfully.

The hwdec looks a bit like the vdpau interop one - the hwdec
maintains a single set of plane textures and each output frame
is repeatedly mapped into these textures to pass on.

The frames are always in NV12 format, at least until 10bit output
supports emerges.

The only slightly interesting part of the copying process is that
CUDA works by associating PBOs, so we need to define these for
each of the textures.

TODO Items:
* There's something weird with interlaced mpeg2 content. It's
  appearing distorted, as if there's a stride problem. Interlaced
  h.264 is fine, so I'm not sure whats up. It might be due to
  fields being passed separately vs packed together.
* I need to add a download_image function for screenshots. This
  would do the same copy to system memory that the decoder's
  system memory output does.

Final note: I mentioned HEVC Main10. While there is no 10bit output
support, NvDecode can return dithered 8bit NV12 so you can take
advantage of the hardware acceleration.

This particular mode requires compiling ffmpeg with a modified
header (or possibly the CUDA 8 RC) and is not upstream in ffmpeg
yet.

Usage:

It's quite verbose, due to needing to specify both the decoder
and the hwdec. Note that hwdec=auto will probably not work as it
will try to use vdpau and the decoder must always be specified as
the default decoder always has higher precedence.

mpv --hwdec=cuda --vd=lavc:h264_cuvid --vo=opengl foobar.mp4

haasn commented:
Something I just realized: This might also slightly improve performance over VDPAU because the latter delivers frames as fields instead of continuous textures (requiring us to weave it back together in GLSL), and by the sound of it CUVID just gives us a single NV12 texture as-is.

philipl commented:
It's a possibility, but I suspect it would be negligible or non-existent. At the lowest level, all the hardware decoders except hevc return fields and the higher level code and/or libraries weave them together in some configurations and not in others.

The hevc situation is amusing because vdpau doesn't fully account for the data being in a single frame and not two fields. If you try and use the direct-mode in mpv with hevc content, you'll get a mess as the frame is weaved when it shouldn't be (top half weaved with bottom half). The first driver release with vdpau support was unusable for hevc because it would run the internal weave on hevc unconditionally and return weaved garbage. They fixed this in the mixer but raw data access is still messed up, whether you use GL interop or the system memory readback.

haasn commented:
> They fixed this in the mixer but raw data access is still messed up, whether you use GL interop or the system memory readback.

What happens currently when mpv does raw access on VDPAU-decoded HEVC? Would it be as simple as just making the VDPAU weaving conditional?

wm4 commented:
> What happens currently when mpv does raw access on VDPAU-decoded HEVC? Would it be as simple as just making the VDPAU weaving conditional?

It messes up. But I think it works correctly with AMD. This is just nvidia's failure to provide working drivers.

wm4 commented:
Seems generally ok. To make --hwdec force specific decoders, you can initialize the vd_lavc_hwdec.lavc_suffix field with ``_cuda``. This would force ``h264_cuda`` etc. (you can probably also drop the codec IDs during probing).

At one point, mpv should probably have proper support for the hwframes stuff, which could probably forego some boilerplate, and automatically use the hwframes copy functions for screenshots, but as long as we want to support ffmpeg releases without hwframes stuff, this is probably not feasible, and it has to be done manually. The frames ctx is still available via mp_image.hwctx, though.

philipl commented:
> It messes up. But I think it works correctly with AMD. This is just nvidia's failure to provide working drivers.

Yeah, we would need to both detect hevc and the nvidia driver to be able to skip direct mode.

philipl commented:
Updated PR with use of lavc_suffix and added #ifdef guards to fix the build.

wm4 commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -0,0 +1,287 @@
+    ret = CHECK_CU(cuDeviceGet(&device, 0));
+    if (ret < 0) {
+        goto error;
+    }
```
Before I forgot it: if the body of an if statement is only one physical line, the `{`/`}` can be left away.

wm4 commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    ret = CHECK_CU(cuMemcpy(cuda_data, (CUdeviceptr)hw_image->planes[0],
+                            hw_image->w * hw_image->h));
```
Are you sure this is the correct stride?

Also in theory you could use a dummy mp_image and use its meta information to handle Y and UV planes in a uniform way, e.g. like the vaapi code: https://github.com/mpv-player/mpv/blob/master/video/out/opengl/hwdec_vaegl.c#L307

philipl replied:
I'm not sure - it's probably the cause of the mpeg2 problem. Thanks for the tip.

wm4 commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -0,0 +1,287 @@
+
+    gl->GenBuffers(2, p->gl_pbos);
+    for (int n = 0; n < 2; n++) {
+        gl->BindBuffer(GL_PIXEL_UNPACK_BUFFER_ARB, p->gl_pbos[n]);
```
Other code in mpv doesn't use the `_ARB` suffix for PBOs.

Also I think you should check for desktop OpenGL in the init function, just to be sure (since this could also run with an old nvidia driver in ES2 mode if the user forces it). This can be done with `if(gl->version>=210)` or so.

philipl commented:
Updated to reflect feedback:

* Added version check
* Removed _ARB suffix
* Used a dummy image to calculate dimensions and refactored code to take advantage of it. It seems my old dimensions were correct. No change to the mpeg2 problem. I'll investigate that after this change is done.

philipl commented:
FYI: The mpeg2 problem appears to be in ffmpeg (or lower). If I do a transcode job using the decoder, I see the same garbled output. Strangely, if I use the cuvid decoder without the hwdec (so doing copies to system memory), it plays back correctly. Very strange.

philipl commented:
Updated the commit description.

haasn commented:
Tested it, seems to work fine on my end (although actually installing CUDA was a challenge).

In case anybody else gets stuck, https://github.com/haasn/gentoo-conf/commit/d3ae2ebdc5e5edf114371d525f385391310cfeef is what I ended up doing.

haasn commented:
Btw, it completely explodes on 10-bit H.264 files: broken video, renders at 1 fps

wm4 commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -0,0 +1,272 @@
+    for (int n = 0; n < 2; n++) {
+        p->gl_textures[n] = 0;
+        p->gl_pbos[n] = 0;
+    }
```
You don't need to reset the fields, because the context is deallocated after `destroy`.

wm4 commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    *out_frame = (struct gl_hwdec_frame){
+    };
```
This is not valid C. You need a 0, e.g. `*out_frame = (struct gl_hwdec_frame){0};`.

philipl commented:
> Btw, it completely explodes on 10-bit H.264 files: broken video, renders at 1 fps
> 
> Edit: It also completely explodes on this otherwise reasonable 8-bit H.264 file: https://github.com/haasn/cms/raw/master/arrietty.mkv

I think your cuda/driver/hardware might be old? At least for me, 10bit H.264 results in a stream of error messages and no frames ever returned. Not great but better, I guess.

And your sample plays fine.

I'm using a 1080 with cuda 7.5 sdk and video sdk 7.0.

philipl commented:
Updated to fix new feedback items.

I'm not doing anything about unsupported profiles here. I think that should be handled on the ffmpeg side.

haasn commented:
> I think your cuda/driver/hardware might be old? At least for me, 10bit H.264 results in a stream of error messages and no frames ever returned. Not great but better, I guess.

GTX 970, nvidia-drivers 370.23, nvidia-cuda-toolkit 7.5.18-r2

philipl commented:
> GTX 970, nvidia-drivers 370.23, nvidia-cuda-toolkit 7.5.18-r2

Don't know what to say. I sold off my Maxwell hardware when I got the 1080 so I can't test. Maybe this stuff only works reliably on Pascal...

philipl commented:
Fixed the alignment issues. My 720x480 mpeg2 sample and an 854x480 h.264 sample now play correctly.

wm4 commented:
Merged.

Kagami commented:
Thank you @philipl for such great contribution!

Just rebuilt ffmpeg and mpv with cuda support (thanks to @haasn for gentoo-specific configuration highlights).

H.264 hardware decoding works as expected. Default: ≈80% load of one CPU core on 1080p file. With `--hwdec=cuda`: ≈5% load.
HEVC Main10 doesn't work as stated in documentation.
Though HEVC Main doesn't work too:

```
[ffmpeg/video] hevc_cuvid: cuvidCreateDecoder(&cudec, &cuinfo) failed ->
CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
Could not open codec.
Using software decoding.
```

Am I doing something wrong?

philipl commented:
@Kagami what hardware are you using? That error sounds like your hardware can't do hevc. You need a GTX 960 (desktop only) or a GTX 10x0 to do it.

Kagami commented:
You're right, I have GTX 970. I thought HEVC decoding via CUVID will work even on Feature Set E cards?

haasn commented:
No, the GTX 970 does not have a HEVC hardware decoder at all.

Kagami commented:
Huh, what a pity. Btw, hwdec=cuda works for me even without `--vd=lavc:h264_cuvid`.

philipl commented:
Yes, there's a subsequent change that adds the suffix configuration so that mpv automatically uses the suffixed decoder.

philipl commented:
For the record, this fixes #735

amietn commented:
I just tried building this on my system with a GTX 1080.

The build went fine and I have tested playing 2 different video files. Both are HEVC, 60 FPS and 3840x2160.

The first file was encoded with Main profile and decoding works just fine.

The second was encoded with Main 10 profile and the result is a green picture like the following: http://i.imgur.com/NllUKRG.jpg

I'm using cuda 7.5, nvidia 370.28 and have built mpv and ffmpeg with the mpv-build scripts and the following options files:

In ```ffmpeg_options```:
```
--enable-cuda
--enable-cuvid
--enable-nonfree
--extra-cflags="-I/opt/cuda/include"
```

and in ```mpv_options```:

```
--enable-cuda-hwaccel
```

Command:
```
mpv/build/mpv --hwdec=cuda --vo=opengl <file>
```

I have also tried playing the same file with the NvDecodeGL sample provided with the NVIDIA Video Codec SDK 7.0.1 and got similar garbled output.


@philipl  I guess the issue is the missing modified ffmpeg header or the use of cuda 8 rc. Could you clarify what I'm missing to play HEVC Main 10? And thanks for the contribution, this is much appreciated!

Kagami commented:
@amietn see discussion here: #3504

Kagami commented:
What about vp9_cuvid? [This table](https://developer.nvidia.com/nvidia-video-codec-sdk#CodecSupportMatrix) states VP9 hardware decoding should work on any GM20X, but still fails with same error for me on GTX970. Could anyone with Maxwell GPU test it?

* Test file: https://a.pomf.cat/nrwxnk.webm
* Test command: `ffmpeg -hide_banner -c:v vp9_cuvid -i v.webm -map v -f null -`
* Test output:

```
[vp9_cuvid @ 0x9bd360] cuvidCreateDecoder(&cudec, &cuinfo) failed -> CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
[matroska,webm @ 0x9bae40] Failed to open codec in av_find_stream_info
[vp9_cuvid @ 0x9bd360] cuvidCreateDecoder(&cudec, &cuinfo) failed -> CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
Input #0, matroska,webm, from 'v.webm':
  Metadata:
    encoder         : Lavf57.2.100
  Duration: 00:00:10.00, start: 0.000000, bitrate: 95 kb/s
    Stream #0:0: Video: vp9, nv12, 1280x720, SAR 1:1 DAR 16:9, 1 fps, 1 tbr, 1k tbn, 1k tbc (default)
[vp9_cuvid @ 0xb77ca0] cuvidCreateDecoder(&cudec, &cuinfo) failed -> CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
Output #0, null, to 'pipe:':
  Metadata:
    encoder         : Lavf57.2.100
    Stream #0:0: Unknown: none, SAR 1:1 DAR 0:0 (default)
    Metadata:
      encoder         : Lavc57.55.101 wrapped_avframe
Stream mapping:
  Stream #0:0 -> #0:0 (vp9 (vp9_cuvid) -> wrapped_avframe (native))
Error while opening decoder for input stream #0:0 : Generic error in an external library
```

philipl commented:
It's unclear. It's possible this only landed in the same silicon as hevc (which means GM206 only) and the table is wrong. I was never in a position to test this when I had Maxwell hardware.

I suspect the table is wrong, given that both hevc and vp9 appeared for the first time in the same tegra hardware release (and the decoder silicon is shared).
